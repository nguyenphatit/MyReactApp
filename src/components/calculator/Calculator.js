import React, { Component } from 'react';
import './Calculator.scss';

class AutoScalingText extends Component {
    state = {
        scale: 1
    };

    componentDidUpdate() {
        const { scale } = this.state

        // get div current
        const node = this.node

        // get parent div current
        const parentNode = node.parentNode

        // width div .calculator__display .auto-scaling-text = 306
        const availableWidth = parentNode.offsetWidth

        const actualWidth = node.offsetWidth
        const actualScale = availableWidth / actualWidth

        if (scale === actualScale)
            return

        if (actualScale < 1) {
            this.setState({ scale: actualScale })
        } else if (scale < 1) {
            this.setState({ scale: 1 })
        }
    }

    render() {
        const { scale } = this.state

        return (
            <div
                className="auto-scaling-text"
                style={{ transform: `scale(${scale},${scale})` }}
                ref={node => this.node = node}
            >{this.props.children}</div>
        )
    }
}


class CalculatorDisplay extends Component {
    render() {
        const { value, ...props } = this.props

        const language = navigator.language || 'vi_VI'
        let formattedValue = parseFloat(value).toLocaleString(language, {
            useGrouping: true,
            maximumFractionDigits: 9
        })

        // Add back missing .0 in e.g. 12.0
        const match = value.match(/\.\d*?(0*)$/)

        if (match)
            formattedValue += (/[1-9]/).test(match[0]) ? match[1] : match[0]

        return (
            <div {...props} className="calculator__display auto-scaling-text">
                <AutoScalingText>{formattedValue}</AutoScalingText>
            </div>
        )
    }
}

class Calculator extends Component {
    constructor(props) {
        super(props);
        this.state = {
            displayValue: '0',
            waitingOperand: false,
            operator: null,
            value: null,
        }
    }

    inputDigit(digit) {
        const { displayValue, waitingOperand } = this.state;
        if (waitingOperand) {
            this.setState({
                displayValue: String(digit),
                waitingOperand: false
            })
        } else {
            this.setState({
                displayValue: displayValue === '0' ? String(digit) : displayValue + digit
            });
        }
    }

    inputDot() {
        const { displayValue, waitingOperand } = this.state;
        if (waitingOperand) {
            this.setState({
                displayValue: '.',
                waitingOperand: false
            });
        } else if (displayValue.indexOf('.') === -1) {
            this.setState({
                displayValue: displayValue + '.',
                waitingOperand: false
            });
        }
    }

    inputClear() {
        this.setState({
            displayValue: '0'
        })
    }

    toggleSign() {
        const { displayValue } = this.state;
        this.setState({
            displayValue: displayValue.charAt(0) === '-' ? displayValue.substr(1) : '-' + displayValue
        })
    }

    inputPercent() {
        const { displayValue } = this.state;
        const value = parseFloat(displayValue) / 100

        this.setState({
            displayValue: String(value)
        })
    }

    operation(nextOperator) {
        const { displayValue, operator, value } = this.state;

        const inputValue = parseFloat(displayValue);

        const operations = {
            '/': (prevValue, nextValue) => prevValue / nextValue,
            '*': (prevValue, nextValue) => prevValue * nextValue,
            '-': (prevValue, nextValue) => prevValue - nextValue,
            '+': (prevValue, nextValue) => prevValue + nextValue,
            '=': (prevValue, nextValue) => nextValue,
        }

        if (value == null) {
            this.setState({
                value: inputValue
            })
        } else if (operator) {
            const currentValue = value || 0;
            const computedValue = operations[operator](currentValue, inputValue);

            this.setState({
                value: computedValue,
                displayValue: String(computedValue)
            })
        }

        this.setState({
            waitingOperand: true,
            operator: nextOperator
        })
    }


    calculatorDisplay(value) {
        const language = navigator.language || 'vi-VI'
        let formattedValue = parseFloat(value).toLocaleString(language, {
            useGrouping: true,
            maximumFractionDigits: 10
        })

        const match = value.match(/\.\d*?(0*)$/)
        if (match)
            formattedValue += (/[1-9]/).test(match[0]) ? match[1] : match[0]
        return formattedValue
    }

    render() {
        const { displayValue } = this.state;
        return (
            <div>
                <div className="calculator">
                    {/* <div className="calculator__display auto-scaling-text">{displayValue.length < 11 ? displayValue : <span className="message_error"># max chars (X_X)!</span>}</div> */}
                    {/* <div className="calculator__display auto-scaling-text"> */}
                    <CalculatorDisplay value={displayValue} />
                    {/* </div> */}
                    <table>
                        <tbody>
                            <tr>
                                <td>
                                    <button className="calculator__key calculator__func" onClick={() => this.inputClear()}>AC</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__func" onClick={() => this.toggleSign()}>±</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__func" onClick={() => this.inputPercent()}>%</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__oper" onClick={() => this.operation('/')}>÷</button>
                                </td>
                            </tr>
                            <tr>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(7)}>7</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(8)}>8</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(9)}>9</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__oper" onClick={() => this.operation('*')}>x</button>
                                </td>
                            </tr>
                            <tr>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(4)}>4</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(5)}>5</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(6)}>6</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__oper" onClick={() => this.operation('-')}>-</button>
                                </td>
                            </tr>
                            <tr>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(1)}>1</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(2)}>2</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDigit(3)}>3</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__oper" onClick={() => this.operation('+')}>+</button>
                                </td>
                            </tr>
                            <tr>
                                <td colSpan="2">
                                    <button className="calculator__key calculator__key0" onClick={() => this.inputDigit(0)}>0</button>
                                </td>
                                <td>
                                    <button className="calculator__key" onClick={() => this.inputDot()}>.</button>
                                </td>
                                <td>
                                    <button className="calculator__key calculator__oper" onClick={() => this.operation('=')}>=</button>
                                </td>
                            </tr>
                        </tbody>
                    </table>
                </div>
            </div >
        );
    }
}

export default Calculator;