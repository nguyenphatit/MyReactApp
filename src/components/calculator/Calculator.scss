@import url('https://fonts.googleapis.com/css?family=Roboto:100,400');

.table {
    width: 100%;
    padding-left: 2px;
}

.auto-scaling-text {
    display: inline-block;
}

.calculator__display {
    font-family: 'Roboto', sans-serif;
    font-weight: 100;
    width: 300px;
    color: #fff;
    padding: 15px 3px;
    text-align: right;
    font-size: 50px;
    background: #232323;
    margin-bottom: 50px;
    & .auto-scaling-text {
        padding: 0 20px;
        position: absolute;
        right: 0;
        transform-origin: right;
    }
    .message_error {
        font-size: 30px;
    }
}

.calculator {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    padding: 0;
    width: 310px;
    background: #232323;
    box-shadow: 0px 0px 15px rgba(0, 0, 0, 0.7);
    &__key {
        font-family: 'Roboto', sans-serif;
        font-weight: 100;
        font-size: 20px;
        padding: 20px 5px;
        min-width: 73px;
        border: 0;
        border-radius: 3px;
        background-color: #cccccc;
        background-image: linear-gradient(#cccccc, rgb(133, 133, 133));
        cursor: pointer;
        &:hover {
            background-color: #cccccc - 10;
        }
        &:active {
            box-shadow: inset 0px 0px 80px 0px rgba(0,0,0,0.25);
        }
    }
    &__key0 {
        min-width: 150px;
    }
    &__func {
        background-color: #999;
        background-image: linear-gradient(#999, rgb(95, 95, 95));
        &:hover {
            background-color: #999 - 30;
        }
    }
    &__oper {
        background-color: orange;
        background-image: linear-gradient(rgb(255, 213, 124), orange);
        &:hover {
            background-color: orange - 30;
            background-image: linear-gradient(rgb(218, 161, 40), rgb(226, 147, 0));
        }
    }
}