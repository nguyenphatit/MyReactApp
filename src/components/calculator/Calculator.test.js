import React from 'react';
import { shallow } from 'enzyme';
import Calculator from './Calculator';
import renderer from 'react-test-renderer';

describe('Calculator Component', () => {
    it('matches the snapshot', () => {
        const tree = renderer.create(<Calculator />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('state displayValue', () => {
        const wrapper = shallow(<Calculator />);
        const resultState = wrapper.state().displayValue;
        expect(resultState).toEqual('0')
    })

    it('state waitingOperand', () => {
        const wrapper = shallow(<Calculator />);
        const resultState = wrapper.state().waitingOperand;
        expect(resultState).toEqual(false)
    })

    it('state operator', () => {
        const wrapper = shallow(<Calculator />);
        const resultState = wrapper.state().displayValue;
        expect(resultState).toEqual('0')
    })

    it('state value', () => {
        const wrapper = shallow(<Calculator />);
        const resultState = wrapper.state().displayValue;
        expect(resultState).toEqual('0')
    })
})