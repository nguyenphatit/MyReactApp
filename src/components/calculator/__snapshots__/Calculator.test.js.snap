// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component matches the snapshot 1`] = `
<div
  className="calculator"
>
  <div
    className="calculator__display auto-scaling-text"
  >
    <div
      className="auto-scaling-text"
      style={
        Object {
          "transform": "scale(1,1)",
        }
      }
    >
      0
    </div>
  </div>
  <table>
    <tbody>
      <tr>
        <td>
          <button
            className="calculator__key calculator__func"
            onClick={[Function]}
          >
            AC
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__func"
            onClick={[Function]}
          >
            ±
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__func"
            onClick={[Function]}
          >
            %
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__oper"
            onClick={[Function]}
          >
            ÷
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            7
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            8
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            9
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__oper"
            onClick={[Function]}
          >
            x
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            4
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            5
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            6
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__oper"
            onClick={[Function]}
          >
            -
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            1
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            2
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            3
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__oper"
            onClick={[Function]}
          >
            +
          </button>
        </td>
      </tr>
      <tr>
        <td
          colSpan="2"
        >
          <button
            className="calculator__key calculator__key0"
            onClick={[Function]}
          >
            0
          </button>
        </td>
        <td>
          <button
            className="calculator__key"
            onClick={[Function]}
          >
            .
          </button>
        </td>
        <td>
          <button
            className="calculator__key calculator__oper"
            onClick={[Function]}
          >
            =
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
